### PR TITLE
Remove linux getrandom and use C++ stdlib instead

### LINKED ISSUE #1249: fatal error: sys/random.h
CC       getrandom_linux.lo
getrandom_linux.c:40:24: fatal error: sys/random.h: No such file or directory
 #include <sys/random.h>
                        ^
compilation terminated.

### LINKED ISSUE #1304: Has error occurred in compile the Aria2c
```shell
autoreconf -i &&./configure &&make
...
  CXX      InternalDHKeyExchange.lo
In file included from InternalDHKeyExchange.h:40:0,
                 from InternalDHKeyExchange.cc:36:
bignum.h: In constructor ‘bignum::ulong<dim>::ulong()’:
bignum.h:34:25: error: ‘make_unique’ was not declared in this scope
   inline ulong() : buf_(make_unique<char_t[]>(dim)) {}
                         ^~~~~~~~~~~
bignum.h:34:25: note: suggested alternative:
In file included from /usr/include/c++/6/memory:81:0,
                 from bignum.h:19,
                 from InternalDHKeyExchange.h:40,
                 from InternalDHKeyExchange.cc:36:
/usr/include/c++/6/bits/unique_ptr.h:802:5: note:   ‘std::make_unique’
     make_unique(_Args&&...) = delete;
     ^~~~~~~~~~~
In file included from InternalDHKeyExchange.h:40:0,
                 from InternalDHKeyExchange.cc:36:
bignum.h:34:43: error: expected primary-expression before ‘[’ token
   inline ulong() : buf_(make_unique<char_t[]>(dim)) {}
                                           ^
bignum.h:34:44: error: expected primary-expression before ‘]’ token
   inline ulong() : buf_(make_unique<char_t[]>(dim)) {}
                                            ^
bignum.h: In constructor ‘bignum::ulong<dim>::ulong(size_t)’:
bignum.h:35:33: error: ‘make_unique’ was not declared in this scope
   inline ulong(size_t t) : buf_(make_unique<char_t[]>(dim))
                                 ^~~~~~~~~~~
bignum.h:35:33: note: suggested alternative:
In file included from /usr/include/c++/6/memory:81:0,
                 from bignum.h:19,
                 from InternalDHKeyExchange.h:40,
                 from InternalDHKeyExchange.cc:36:
/usr/include/c++/6/bits/unique_ptr.h:802:5: note:   ‘std::make_unique’
     make_unique(_Args&&...) = delete;
     ^~~~~~~~~~~
In file included from InternalDHKeyExchange.h:40:0,
                 from InternalDHKeyExchange.cc:36:
bignum.h:35:51: error: expected primary-expression before ‘[’ token
   inline ulong(size_t t) : buf_(make_unique<char_t[]>(dim))
                                                   ^
bignum.h:35:52: error: expected primary-expression before ‘]’ token
   inline ulong(size_t t) : buf_(make_unique<char_t[]>(dim))
                                                    ^
bignum.h: In copy constructor ‘bignum::ulong<dim>::ulong(const bignum::ulong<dim>&)’:
bignum.h:39:46: error: ‘make_unique’ was not declared in this scope
   inline ulong(const ulong<dim>& rhs) : buf_(make_unique<char_t[]>(dim))
                                              ^~~~~~~~~~~
bignum.h:39:46: note: suggested alternative:
In file included from /usr/include/c++/6/memory:81:0,
                 from bignum.h:19,
                 from InternalDHKeyExchange.h:40,
                 from InternalDHKeyExchange.cc:36:
/usr/include/c++/6/bits/unique_ptr.h:802:5: note:   ‘std::make_unique’
     make_unique(_Args&&...) = delete;
     ^~~~~~~~~~~
In file included from InternalDHKeyExchange.h:40:0,
                 from InternalDHKeyExchange.cc:36:
bignum.h:39:64: error: expected primary-expression before ‘[’ token
   inline ulong(const ulong<dim>& rhs) : buf_(make_unique<char_t[]>(dim))
                                                                ^
bignum.h:39:65: error: expected primary-expression before ‘]’ token
   inline ulong(const ulong<dim>& rhs) : buf_(make_unique<char_t[]>(dim))
                                                                 ^
bignum.h: In constructor ‘bignum::ulong<dim>::ulong(const char_t*, size_t)’:
bignum.h:44:14: error: ‘make_unique’ was not declared in this scope
       : buf_(make_unique<char_t[]>(dim))
              ^~~~~~~~~~~
bignum.h:44:14: note: suggested alternative:
In file included from /usr/include/c++/6/memory:81:0,
                 from bignum.h:19,
                 from InternalDHKeyExchange.h:40,
                 from InternalDHKeyExchange.cc:36:
/usr/include/c++/6/bits/unique_ptr.h:802:5: note:   ‘std::make_unique’
     make_unique(_Args&&...) = delete;
     ^~~~~~~~~~~
In file included from InternalDHKeyExchange.h:40:0,
                 from InternalDHKeyExchange.cc:36:
bignum.h:44:32: error: expected primary-expression before ‘[’ token
       : buf_(make_unique<char_t[]>(dim))
                                ^
bignum.h:44:33: error: expected primary-expression before ‘]’ token
       : buf_(make_unique<char_t[]>(dim))
                                 ^
Makefile:2300: recipe for target 'InternalDHKeyExchange.lo' failed
make[3]: *** [InternalDHKeyExchange.lo] Error 1
make[3]: Leaving directory '/root/aria2-1.34.0/src'
Makefile:2341: recipe for target 'all-recursive' failed
make[2]: *** [all-recursive] Error 1
make[2]: Leaving directory '/root/aria2-1.34.0/src'
Makefile:556: recipe for target 'all-recursive' failed
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory '/root/aria2-1.34.0'
Makefile:467: recipe for target 'all' failed
make: *** [all] Error 2

```

Platform:Raspberrypi 2B
Version:1.34.0

Why it happen?

### PR DESCRIPTION
**Pull request by tatsuhiro-t**
Fixes #1304 
Fixes #1249

### CODE REVIEW

**Comment by hartwork** [comment]
@tatsuhiro-t are there any guarantees that `std::uniform_int_distribution` will produce as good of any entropy as `getrandom`? [This post hints otherwise](http://anadoxin.org/blog/c-shooting-yourself-in-the-foot-4.html). If plain insecure `rand` was not good enough for aria2, I'd expect need for good entropy. Can you summarize what aria2 needs good entropy for or why bad (or not guaranteed to be good) entropy would not be a security concern in your eyes? Thanks!

**Comment by tatsuhiro-t** [comment]
The counterpart of getrandom in C++1x API is std::random_device.  The difference of getrandom and the current code is whether sampling from /dev/urandom or use pseudo random generator mt19937.
This function is used when we need some random data like producing ID or padding length.
For security stuff, we use the function provided by crypto library (e.g., OpenSSL).
The one exception is BitTorrent InternalDHEKeyExchange which uses this random generator to produce keys.  It is only used if no cyrpto library is available.  So it might be affected but the current code has existed as fallback so we didn't lower the bar.  Also BitTorrent encryption is considered as insecure.

**Comment by hartwork** [comment]
That was quick! I read your reply as "we don't need high quality entropy for all cases that involve this function" — is that understanding correct?

**Comment by tatsuhiro-t** [comment]
yes
